src/neuroglancer/viewer.ts

```diff
@@ -773,32 +812,69 @@ export class Viewer extends RefCounted implements ViewerState {
     }
   }
 
-  postJsonState() {
-    // if jsonStateServer is not present prompt for value and store it in state
-    if (!this.jsonStateServer.value) {
-      this.promptJsonStateServer(
-          'No state server found. Please enter a server URL, or hit OK to use the default server.');
+  loadFromJsonUrl() {
+    var urlParams = new URLSearchParams(window.location.search);
+    if (urlParams.has('json_url')) {
+      let json_url = urlParams.get('json_url')!;
+      history.replaceState(null, '', removeParameterFromUrl(window.location.href, 'json_url'));
+      this.resetStateWhenEmpty = false;
+      StatusMessage
+          .forPromise(
+              cancellableFetchOk(json_url, {}, responseJson).then(response => {
+                this.state.restoreState(response);
+              }),
+              {
+                initialMessage: `Retrieving state from json_url: ${json_url}.`,
+                delay: true,
+                errorPrefix: `Error retrieving state: `,
+              })
+          .finally(() => {
+            this.resetStateWhenEmpty = true;
+          });
     }
+  }
 
+  postJsonState(get: UrlType = 0) {
     // upload state to jsonStateServer (only if it's defined)
-    if (this.jsonStateServer.value) {
+    if (this.saver && this.saver.activeEntry && !this.saver.activeEntry.dirty) {
+      this.showSaveDialog(get, this.saver.savedUrl);
+      return;
+    }
+    if (this.jsonStateServer.value || !get) {
       StatusMessage.showTemporaryMessage(`Posting state to ${this.jsonStateServer.value}.`);
       cancellableFetchOk(
           this.jsonStateServer.value, {method: 'POST', body: JSON.stringify(this.state.toJSON())},
           responseJson)
           .then(response => {
-            history.replaceState(
-                null, '',
-                window.location.origin + window.location.pathname + '?json_url=' + response);
+            const savedUrl =
+                `${window.location.origin}${window.location.pathname}?json_url=${response}`;
+            if (this.saver && this.saver.supported) {
+              this.saver.commit(response);
+              this.showSaveDialog(get, response);
+            } else {
+              history.replaceState(null, '', savedUrl);
+              this.showSaveDialog(get);
+            }
           })
           // catch errors with upload and prompt the user if there was an error
           .catch(() => {
             this.promptJsonStateServer('state server not responding, enter a new one?');
             if (this.jsonStateServer.value) {
               this.postJsonState();
             }
+          })
+          .finally(() => {
+            const noStateServerAccess = !this.jsonStateServer.value;
+            if (noStateServerAccess) {
+              this.hashBinding!.legacy.setUrlHash();
```

Comment:
`hashBinding` _could_ be undefined, right? I thought it's only initialized if local storage does not work?

Reply:
added a comment explaining. hashBinding is not optional hashBinding.legacy.falback is